src/parser/parser.lisp

```diff
@@ -0,0 +1,1491 @@
+  (vars   (util:required 'vars)   :type identifier-src-list :read-only t)
+  (body   (util:required 'body)   :type node-body           :read-only t)
+  (source (util:required 'source) :type cons                :read-only t))
+
+(eval-when (:load-toplevel :compile-toplevel :execute)
+  (defun instance-method-definition-list-p (x)
+    (and (alexandria:proper-list-p x)
+         (every #'instance-method-definition-p x))))
+
+(deftype instance-method-definition-list ()
+  '(satisfies instance-method-definition-list-p))
+
+(defstruct (toplevel-define-instance
+            (:copier nil))
+  (context  (util:required 'context)  :type ty-predicate-list               :read-only t)
+  (pred     (util:required 'pred)     :type ty-predicate                    :read-only t)
+  (methods  (util:required 'methods)  :type instance-method-definition-list :read-only t)
+  (source   (util:required 'source)   :type cons                            :read-only t)
+  ;; Source information for the context and the pred
+  (head-src (util:required 'head-src) :type cons                            :read-only t))
+
+(eval-when (:load-toplevel :compile-toplevel :execute)
+  (defun toplevel-define-instance-list-p (x)
+    (and (alexandria:proper-list-p x)
+         (every #'toplevel-define-instance-p x))))
+
+(deftype toplevel-define-instance-list ()
+  '(satisfies toplevel-define-instance-list-p))
+
+(defstruct (program
+            (:copier nil))
+  (package   (util:required 'package) :type package                       :read-only t)
+  (file      (util:required 'file)    :type sb-sys::fd-stream             :read-only t)
+  (types     nil                      :type toplevel-define-type-list     :read-only nil)
+  (declares  nil                      :type toplevel-declare-list         :read-only nil)
+  (defines   nil                      :type toplevel-define-list          :read-only nil)
+  (classes   nil                      :type toplevel-define-class-list    :read-only nil)
+  (instances nil                      :type toplevel-define-instance-list :read-only nil))
+
+;;
+;; Empty package for reading (package) forms
+;;
+(defpackage #:coalton-impl/parser/read
+  (:use))
+
+;; TODO: handle trailing attributes
+
+(defun parse-file (filename)
+  (with-open-file (file filename :if-does-not-exist :error)
+    (let* (;; Setup eclector readtable
+           (eclector.readtable:*readtable*
+             (eclector.readtable:copy-readtable eclector.readtable:*readtable*))
+
+           ;; Initial package to read (package) forms into
+           (*package* (find-package "COALTON-IMPL/PARSER/READ"))
+
+           ;; Read unspecified floats as double floats
+           (*read-default-float-format* 'double-float)
+
+           ;; Read the (package) form
+           (package-form (eclector.concrete-syntax-tree:read file nil 'eof)))
+
+      (when (eq package-form 'eof)
+        (error "file starts with eof"))
+
+      (setf *package* (parse-package package-form file))
```

Comment:
is there a reason to not bind `*package*` instead of setting it for the context of this function?